R/coerce.R

```diff
@@ -0,0 +1,165 @@
+#'   sources.
+#' @param image_id Optional character vector (same length as
+#'   \code{imageSources}) containing unique image identifiers.
+#' @param loadImage Logical indicating whether to load image into memory.
+#'   Default = \code{FALSE}.
+
+#' @aliases 
+#' coerce, SingleCellExperiment, SpatialExperiment-method
+#' toSpatialExperiment
+#' 
+#' @examples 
+#' dir <- system.file(
+#' file.path("extdata", "10xVisium", "section1"),
+#' package = "SpatialExperiment")
+#' 
+#' # read in counts
+#' fnm <- file.path(dir, "raw_feature_bc_matrix")
+#' sce <- DropletUtils::read10xCounts(fnm)
+#' 
+#' 
+#' # read in spatial coordinates
+#' fnm <- file.path(dir, "spatial", "tissue_positions_list.csv")
+#' xyz <- read.csv(fnm, header = FALSE,
+#'     col.names = c("barcode", "in_tissue", "array_row", "array_col",
+#'     "pxl_row_in_fullres", "pxl_col_in_fullres"))
+#' ## as method
+#' (spe <- as(sce, "SpatialExperiment"))
+#' int_colData(sce)$spatialData <- DataFrame(xyz[,c(1:4)])
+#' int_colData(sce)$spatialCoords <- as.matrix(xyz[,c(5,6)])
+#' ## Coercing an sce without imgData
+#' (spe <- as(sce, "SpatialExperiment"))
+#' ## Coercing an sce with imgData
+#' img <- readImgData(
+#'     path = file.path(dir, "spatial"),
+#'     sample_id = "sample01")
+#' int_colData(sce)$imgData <- img
+#' (spe <- as(sce, "SpatialExperiment"))
+#' 
+#' ## toSpatialExperiment method
+#' (spe <- toSpatialExperiment(sce,
+#'     imgData = img,
+#'     spatialData = DataFrame(xyz), 
+#'     spatialCoordsNames = c("pxl_col_in_fullres", "pxl_row_in_fullres"),
+#'     sample_id = "sample01"))
+NULL
+
+
+setAs(
+    from="SingleCellExperiment", 
+    to="SpatialExperiment", 
+    function(from) {
+        spatialCoords <- spatialData <- imgData <- NULL
+        sample_id <- unique(from$sample_id)
+        if (is.null(sample_id)) sample_id <- "sample01"
+        icd <- int_colData(from)
+        spe <- .sce_to_spe(from, 
+            sample_id=sample_id,
+            spatialData=icd$spatialData,
+            spatialCoords=icd$spatialCoords,
+            imgData=icd$imgData)
+        return(spe)
+    }
+)
+
+#' @export
+toSpatialExperiment <- function(sce,
+                                sample_id="sample01", 
+                                spatialDataNames=NULL,
+                                spatialCoordsNames=NULL,
+                                spatialData=NULL,
+                                spatialCoords=NULL,
+                                scaleFactors=1,
+                                imageSources=NULL,
+                                image_id=NULL,
+                                loadImage=TRUE,
+                                imgData=NULL) {
+    
+    stopifnot(is(sce, "SingleCellExperiment"))
+    
+    ## giving priority to passed arguments
+    if (all(is.null(spatialCoords), is.null(spatialCoordsNames))) {
+        spatialCoords <- int_colData(sce)$spatialCoords
+        spatialCoordsNames <- names(spatialCoords)
```

Comment:
This should be `colnames(spatialCoords)`, as `spatialCoords` should be a matrix (`names` will return NULL!).

Reply:
good point!